File: book/petrol-pump/chapter3/section7/Keypad.java
package chapter3.section7;

import pump.*;
import sodium.*;
import java.util.Optional;

public class Keypad {
    public final Behavior<Integer> value;
    public final Event<Unit> eBeep;

    public Keypad(Event<Key> eKeypad,
                  Event<Unit> eClear,
                  Behavior<Boolean> active) {
        this(eKeypad.gate(active), eClear);
    }

    public Keypad(Event<Key> eKeypad, Event<Unit> eClear) {
        BehaviorLoop<Integer> value = new BehaviorLoop<>();
        this.value = value;
        Event<Integer> eKeyUpdate = Event.filterOptional(
            eKeypad.snapshot(value,
                (key, value_) -> {
                    if (key == Key.CLEAR)
                        return Optional.of(0);
                    else {
                        int x10 = value_ * 10;
                        return x10 >= 1000
                                ? Optional.empty()
                                : Optional.of(
                                    key == Key.ZERO ? x10 :
                                    key == Key.ONE ? x10 + 1 :
                                    key == Key.TWO ? x10 + 2 :
                                    key == Key.THREE ? x10 + 3 : 
                                    key == Key.FOUR ? x10 + 4 : 
                                    key == Key.FIVE ? x10 + 5 : 
                                    key == Key.SIX ? x10 + 6 : 
                                    key == Key.SEVEN ? x10 + 7 : 
                                    key == Key.EIGHT ? x10 + 8 : 
                                                       x10 + 9
                                );
                    }
                }
            )
        );
        value.loop(eKeyUpdate.merge(eClear.map(u -> 0))
                             .hold(0));
        eBeep = eKeyUpdate.map(k -> Unit.UNIT);
    }
}

File: c#/Book/PetrolPump/PetrolPump/PetrolPump/Chapter3/Section8/KeyPad.cs
using System;

using PetrolPump.Pump;

using Sodium;

namespace PetrolPump.Chapter3.Section8
{
  public class Keypad
  {
    public readonly Behavior<int> Value;
    public readonly Event<Unit> EBeep;

    public Keypad(
      Event<Key> eKeypad,
      Event<Unit> eClear,
      Behavior<Boolean> active)
      : this(eKeypad.Gate(active), eClear)
    {}

    public Keypad(Event<Key> eKeypad, Event<Unit> eClear)
    {
      BehaviorLoop<int> value = new BehaviorLoop<int>();
      this.Value = value;
      Event<int> eKeyUpdate = Event<int>.FilterOptional(
        eKeypad.Snapshot(
          value,
          (key, value_) =>
          {
            if (key == Key.CLEAR)
              return Optional<int>.Of(0);
            else
            {
              int x10 = value_ * 10;
              return x10 >= 1000
                ? Optional<int>.Empty()
                : Optional<int>.Of(
                  key == Key.ZERO
                    ? x10
                    : key == Key.ONE
                      ? x10 + 1
                      : key == Key.TWO
                        ? x10 + 2
                        : key == Key.THREE
                          ? x10 + 3
                          : key == Key.FOUR
                            ? x10 + 4
                            : key == Key.FIVE
                              ? x10 + 5
                              : key == Key.SIX
                                ? x10 + 6
                                : key == Key.SEVEN
                                  ? x10 + 7
                                  : key == Key.EIGHT
                                    ? x10 + 8
                                    : x10 + 9
                  );
            }
          })
        );
      value.Loop(
        eKeyUpdate.Merge(eClear.Map(u => 0))
          .Hold(0));
      EBeep = eKeyUpdate.Map(k => Unit.UNIT);
    }
  }

}
